third_party/move/move-compiler-v2/src/pipeline/explicit_drop.rs

```diff
@@ -0,0 +1,197 @@
+                self.drop_temp(arg, attr_id)
+            }
+        }
+    }
+
+    // Returns a set of locals that can be dropped at given code offset
+    // Primitives are filtered out
+    fn released_temps_at(&self, code_offset: CodeOffset) -> BTreeSet<TempIndex> {
+        let live_var_info = self.get_live_var_info(code_offset);
+        let lifetime_info = self.get_lifetime_info(code_offset);
+        let bytecode = &self.target.get_bytecode()[code_offset as usize];
+        released_temps(live_var_info, lifetime_info, bytecode)
+            .into_iter()
+            .filter(|t| !matches!(self.target.get_local_type(*t), Type::Primitive(_)))
+            .collect()
+    }
+
+    fn get_live_var_info(&self, code_offset: CodeOffset) -> &'a LiveVarInfoAtCodeOffset {
+        self.live_var_annot
+            .get_live_var_info_at(code_offset)
+            .expect("live var info")
+    }
+
+    fn get_lifetime_info(&self, code_offset: CodeOffset) -> &'a LifetimeInfoAtCodeOffset {
+        self.lifetime_annot.get_info_at(code_offset)
+    }
+
+    fn drop_temp(&mut self, tmp: TempIndex, attr_id: AttrId) {
+        let drop_t = Bytecode::Call(attr_id, Vec::new(), Operation::Destroy, vec![tmp], None);
+        self.emit_bytecode(drop_t)
+    }
+
+    fn drop_temps(&mut self, temps_to_drop: &BTreeSet<TempIndex>, attr_id: AttrId) {
+        for t in temps_to_drop {
+            self.drop_temp(*t, attr_id)
+        }
+    }
+
+    fn emit_bytecode(&mut self, bytecode: Bytecode) {
+        self.transformed.push(bytecode)
+    }
+}
+
+// Returns a set of locals that can be dropped
+// these are the ones no longer alive or borrowed
+// including locals of primitives
+fn released_temps(
+    live_var_info: &LiveVarInfoAtCodeOffset,
+    life_time_info: &LifetimeInfoAtCodeOffset,
+    bytecode: &Bytecode,
+) -> BTreeSet<TempIndex> {
+    // use set to avoid duplicate dropping
+    let mut released_temps = BTreeSet::new();
+    for t in live_var_info.released_temps() {
+        if !life_time_info.after.is_borrowed(t) {
+            released_temps.insert(t);
+        }
+    }
+    for t in life_time_info.released_temps() {
+        if !live_var_info.after.contains_key(&t) {
+            released_temps.insert(t);
+        }
+    }
+    // if a temp is moved, then no need to drop
+    // this should come before the calculation
+    // of unused vars; because of, for instance,
+    // x = move(x)
+    for moved_src in moved_srcs(bytecode) {
+        released_temps.remove(&moved_src);
+    }
+    // this is needed because unused vars are not released by live var info
+    for dst in bytecode.dests() {
+        if !live_var_info.before.contains_key(&dst)
+            && !live_var_info.after.contains_key(&dst)
+            && !life_time_info.before.is_borrowed(dst)
+            && !life_time_info.after.is_borrowed(dst)
+        {
+            // TODO: triggered in ability-checker/ability_violation.move
```

Comment:
This comment is a bit obscure: maybe clarify?

Reply:
The debug assertion is triggered in the mentioned test case.
The original code is
```rust
    for dst in bytecode.dests() {
        if !live_var_info.before.contains_key(&dst)
            && !live_var_info.after.contains_key(&dst)
            // && !life_time_info.before.is_borrowed(dst) 
            // && !life_time_info.after.is_borrowed(dst)
        {
            debug_assert!(
                !life_time_info.after.is_borrowed(dst),
                "dead assignment borrowed later"
            );
            released_temps.insert(dst);
        }
    }
```

Reply:
If not recorded already in an issue, please file one. Sorry to be pedantic about this, but since there are a bunch of possible todos coming out of this PR, we would be wise to document what needs to be done.

Reply:
@fEst1ck The Live Var Analysis seems to mark a var as used once there are no pending uses.  It may have been borrowed before it was known to be not used.  For example:
```
    let y = ...;
    let r_y = &y;
    if (P) {
        // live here
        use y and/or r_y;
        // dead here (A)
     } else {
        // dead here (B)
    }
    // dead here. (C)
```
In this case, before any DCE is done, at points A, B, and C, `y` is dead but also borrowed until you also drop `r_y`. You will need to drop `r_y` before dropping `y`.  If you can drop `y` and `r_y` at `A` and `B` then we will be happy.

Reply:
https://github.com/aptos-labs/aptos-core/issues/11379